### ISSUE TITLE
Spider Unhandled Error : f.seek(-size-self.SIZE_SIZE, os.SEEK_END) 	exceptions.IOError

### ISSUE THREAD

**Issue by fpassaniti** [scrapy/scrapy#845]
The spider was working correctly until this : 

```
2014-08-08 11:55:01+0200 [seloger] DEBUG: Scraped from <200 http://www.seloger.com/annonces/achat/maison/nanteau-sur-essonne-77/91520225.htm?bd=Detail_Nav&div=2238&idtt=2&idtypebien=all&tri=d_dt_crea>
    {'area': u'Nanteau sur Essonne',
     'id': '91520225',
     'nbroom': u'2',
     'nbsleepingroom': u'1',
     'phone': u'02 38 34 87 48',
     'price': u'88000',
     'source': 'seloger',
     'surface': u'18',
     'text': u"5'de malesherbes, espace et libert\xe9, c'est la sensation que vous \xe9prouverez en visitant ce terrain de loisirs de 2236 m\xb2 avec chalet et garage ferm\xe9, en bordure de rivi\xe8re dans un cadre exceptionnel ! \xc0 saisir ! Classe \xe9nergie: vierge.",
     'title': u'Maison Nanteau Sur Essonne 2 pi\xe8ce (s) 18 m\xb2',
     'type': 'maison',
     'url': u'http://www.seloger.com/annonces/achat/maison/nanteau-sur-essonne-77/91520225.htm?p=CCCPqX4AAKvgYyNg'}
2014-08-08 11:55:01+0200 [-] Unhandled Error
    Traceback (most recent call last):
      File "/Library/Python/2.7/site-packages/scrapy/crawler.py", line 93, in start
        self.start_reactor()
      File "/Library/Python/2.7/site-packages/scrapy/crawler.py", line 130, in start_reactor
        reactor.run(installSignalHandlers=False)  # blocking call
      File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/twisted/internet/base.py", line 1169, in run
        self.mainLoop()
      File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/twisted/internet/base.py", line 1178, in mainLoop
        self.runUntilCurrent()
    --- <exception caught here> ---
      File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/twisted/internet/base.py", line 800, in runUntilCurrent
        call.func(*call.args, **call.kw)
      File "/Library/Python/2.7/site-packages/scrapy/utils/reactor.py", line 41, in __call__
        return self._func(*self._a, **self._kw)
      File "/Library/Python/2.7/site-packages/scrapy/core/engine.py", line 107, in _next_request
        if not self._next_request_from_scheduler(spider):
      File "/Library/Python/2.7/site-packages/scrapy/core/engine.py", line 134, in _next_request_from_scheduler
        request = slot.scheduler.next_request()
      File "/Library/Python/2.7/site-packages/scrapy/core/scheduler.py", line 64, in next_request
        request = self._dqpop()
      File "/Library/Python/2.7/site-packages/scrapy/core/scheduler.py", line 94, in _dqpop
        d = self.dqs.pop()
      File "/Library/Python/2.7/site-packages/queuelib/pqueue.py", line 43, in pop
        m = q.pop()
      File "/Library/Python/2.7/site-packages/scrapy/squeue.py", line 18, in pop
        s = super(SerializableQueue, self).pop()
      File "/Library/Python/2.7/site-packages/queuelib/queue.py", line 157, in pop
        self.f.seek(-size-self.SIZE_SIZE, os.SEEK_END)
    exceptions.IOError: [Errno 22] Invalid argument

2014-08-08 11:55:01+0200 [-] Unhandled Error
    Traceback (most recent call last):
      File "/Library/Python/2.7/site-packages/scrapy/crawler.py", line 93, in start
        self.start_reactor()
      File "/Library/Python/2.7/site-packages/scrapy/crawler.py", line 130, in start_reactor
        reactor.run(installSignalHandlers=False)  # blocking call
      File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/twisted/internet/base.py", line 1169, in run
        self.mainLoop()
      File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/twisted/internet/base.py", line 1178, in mainLoop
        self.runUntilCurrent()
    --- <exception caught here> ---
      File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/twisted/internet/base.py", line 800, in runUntilCurrent
        call.func(*call.args, **call.kw)
      File "/Library/Python/2.7/site-packages/scrapy/utils/reactor.py", line 41, in __call__
        return self._func(*self._a, **self._kw)
      File "/Library/Python/2.7/site-packages/scrapy/core/engine.py", line 107, in _next_request
        if not self._next_request_from_scheduler(spider):
      File "/Library/Python/2.7/site-packages/scrapy/core/engine.py", line 134, in _next_request_from_scheduler
        request = slot.scheduler.next_request()
      File "/Library/Python/2.7/site-packages/scrapy/core/scheduler.py", line 64, in next_request
        request = self._dqpop()
      File "/Library/Python/2.7/site-packages/scrapy/core/scheduler.py", line 94, in _dqpop
        d = self.dqs.pop()
      File "/Library/Python/2.7/site-packages/queuelib/pqueue.py", line 43, in pop
        m = q.pop()
      File "/Library/Python/2.7/site-packages/scrapy/squeue.py", line 18, in pop
        s = super(SerializableQueue, self).pop()
      File "/Library/Python/2.7/site-packages/queuelib/queue.py", line 157, in pop
        self.f.seek(-size-self.SIZE_SIZE, os.SEEK_END)
    exceptions.IOError: [Errno 22] Invalid argument

2014-08-08 11:55:03+0200 [seloger] INFO: Crawled 174 pages (at 23 pages/min), scraped 165 items (at 17 items/min)
2014-08-08 11:56:03+0200 [seloger] INFO: Crawled 174 pages (at 0 pages/min), scraped 165 items (at 0 items/min)
2014-08-08 11:57:03+0200 [seloger] INFO: Crawled 174 pages (at 0 pages/min), scraped 165 items (at 0 items/min)
```

and nothing more... it stopped here.

does someone has ever seen something similar ?

thanks

**Comment by asuraphel** [comment]
Hi all. This also happened to me while the crawler was running! It'd be great atleast if you give us pointers how to fix the bad state. Thanks

**Comment by redapple** [comment]
@asuraphel , what version of scrapy and queuelib did you get the error with?
could you share the traceback if you still have it?
If it's happening with a JOBDIR , using another directory would do a fresh restart. Not a fix though.

**Comment by asuraphel** [comment]
@redapple I'm using v1.0.5. I'm not sure about the qlib version. it came with the scrapy. I solved my problem by doing a fresh start and copying the previous `requests.seen` file to the new one.

**Comment by lazem** [comment]
Hi guys,

Any news about this issue? Doing a fresh start using the previous `requests.seen` file doesn't work for me. The spider closes after few requests. Any pointer would be greatly appreciated.

Thanks

**Comment by redapple** [comment]
@lazem , unfortunately, we don't have a way to reproduce it, so it's hard to investigate. If you find something, do let us know here.

**Comment by lubobill1990** [comment]
I have the same problem with scrapy 1.4.0

```
2018-02-14 11:50:03 [twisted] CRITICAL: Unhandled Error
Traceback (most recent call last):
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/scrapy/commands/crawl.py", line 58, in run
    self.crawler_process.start()
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/scrapy/crawler.py", line 291, in start
    reactor.run(installSignalHandlers=False)  # blocking call
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/twisted/internet/base.py", line 1243, in run
    self.mainLoop()
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/twisted/internet/base.py", line 1252, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/twisted/internet/base.py", line 878, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/scrapy/utils/reactor.py", line 41, in __call__
    return self._func(*self._a, **self._kw)
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/scrapy/core/engine.py", line 122, in _next_request
    if not self._next_request_from_scheduler(spider):
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/scrapy/core/engine.py", line 149, in _next_request_from_scheduler
    request = slot.scheduler.next_request()
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/scrapy/core/scheduler.py", line 71, in next_request
    request = self._dqpop()
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/scrapy/core/scheduler.py", line 106, in _dqpop
    d = self.dqs.pop()
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/queuelib/pqueue.py", line 43, in pop
    m = q.pop()
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/scrapy/squeues.py", line 19, in pop
    s = super(SerializableQueue, self).pop()
  File "/home/lubo/.conda/envs/scrapy/lib/python3.6/site-packages/queuelib/queue.py", line 162, in pop
    self.f.seek(-size-self.SIZE_SIZE, os.SEEK_END)
builtins.OSError: [Errno 22] Invalid argument
````

**Comment by ehrencrona** [comment]
I just experienced this error, too, after running scrapy for about ten hours. I started several instances of the spider running in parallel by mistake at the beginning of the spidering process. Could it be that having multiple scrapys running against the same crawl directory corrupts it?

**Comment by dangra** [comment]
@ehrencrona: Definitively, queuelib is not multiprocesses safe (not even thread safe) so running 2 crawls in parallel for same spider can corrupt it for sure.

**Comment by dangra** [comment]
the only option I see is adding a lockfile solution around queues to prevent corruption due to multi processes trying to update the same file.

**Comment by ibz** [comment]
I have the same issue. Was running just one spider, so it is not a concurrency issue. What happened in my case however is that I ran out of disk space which made the spider crash. It happened once, and restart worked. Then it happened again, and now I am getting this. So it's not always happening, but it should happen *eventually* when running out of disk space. Unfortunately I can't scroll back far enough to see the out of disk space error. I guess that would have helped even more with the investigation. :(

**Comment by SyauqiMA** [comment]
I wonder if this issue already resolved? I have the same problem with this error. 
I ran a spider with JOBDIR=crawl-job-1, after scraping about a thousand pages or so, it's outputting this exact error over and over again.
After deleting the "0" and "1" files in requests.queue and emptying the array at active.json it starts working again, but it seems it starts from the beginning, and after a while it is outputting the same error again.

Is my case related to this particular thread? Or should I make a separate issue?

**Comment by Gallaecio** [comment]
@SyauqiMA If your error is reproducible, and does not involve the scenarios mentioned (running the same spider multiple times in parallel, or running out of disk), feel free to open a separate ticker with a minimal, reproducible example of your scenario (e.g. with a spider pointing to https://toscrape.com/), and include a reference to this ticket just in case.

**Comment by 54keesh** [comment]
I am encountering this issue as well but sadly no way to reciprocate. This is my assumption but i think it might be due to out of scrapper going out of memory as this time i ran my crawler for the maximum time out of all the crawl runs for abt 8 hours or so and my max memory usage was 250 MB on a 1GB memory EC2 instance where 500 or so MB is taken up by the OS.

Fortunately I had a backup of a crawls state directory for a day before and i used that state so only ended up loosing a days worth of crawls.

Using the requests.seen as stated by someone above does seem like a logical solution as it will filter out the dupes upon reprocessing but i guess not all the requests are appended up in requests.seen as my latest requests.seen has only 20k entries while my crawled requests which i am sure are way above 65k including 3x images request per each request and hence this solution didnt work as the metric at the end of job pause did show it filtered dupes but upon inspecting my csvs, it still had dupe entries

**Comment by 54keesh** [comment]
I can confirm that the issue is still there, even after using a previous days crawl state. on some requests the same issue happens but on same it proceeds. It seems like a dead end for now

**Comment by 54keesh** [comment]
@wRAR Why was this issue closed based on PR #7877 only adds another storage engine for the scheduled queues, just like scrapy-redis which work quite efficiently.

I think the actual issue here is; the usage of `JOBDIR` resumable crawls along with no external storage engine so Pickle as a request serializer. This combination causes such an issue. 

We should actually have opt-in support for crash tolerant crawls as issue #7978 mentions but the actual implementation matters where we have to move away from binary data structure in state file (that gets corrupted) and this issue remains as is until that is resolved OR this can be considered as out of scope.
